### PR TITLE
Pullrequest video first implementation

### PR DESCRIPTION
**Pull request by BenSartor**
This is basically the video implementation I did in 2015.
I want it in here for historic reasons ;) and keeps a lot of tryout knowledge about the liblinphone api.
Further commits will follow polishing this up.
See branch [devel-video](../tree/devel-video)

### CODE REVIEW

File: app/src/main/java/org/simlar/service/liblinphone/LinphoneHandler.java [hunk 1]
```diff
@@ -377,4 +383,150 @@ private static void enableDebugMode(final boolean enabled)
...
+		enableCamera(videoPreviewWindow != null);
+		mLinphoneCore.setPreviewWindow(videoPreviewWindow);
+	}
+
+	private void setFrontCameraAsDefault()
+	{
+		int cameraId = 0;
+		for (final AndroidCameraConfiguration.AndroidCamera androidCamera : AndroidCameraConfiguration.retrieveCameras()) {
+			if (androidCamera.frontFacing) {
+				cameraId = androidCamera.id;
```
Comment (jritzerfeld) [inline app/src/main/java/org/simlar/service/liblinphone/LinphoneHandler.java]: Is it correct that to choose choose the **last** found front facing camera?

File: app/src/main/java/org/simlar/service/liblinphone/LinphoneHandler.java [hunk 2]
```diff
@@ -377,4 +383,150 @@ private static void enableDebugMode(final boolean enabled)
...
+			}
+		}
+
+		mLinphoneCore.setVideoDevice(cameraId);
```
Comment (jritzerfeld) [inline app/src/main/java/org/simlar/service/liblinphone/LinphoneHandler.java]: If no camera was found this, I would expect an error.

File: app/src/main/java/org/simlar/service/liblinphone/LinphoneHandler.java [hunk 3]
```diff
@@ -377,4 +383,150 @@ private static void enableDebugMode(final boolean enabled)
...
+		if (currentCall == null) {
+			Lg.w("no current call to toggle camera for");
+			return;
+		}
+
+		final int currentCameraId = mLinphoneCore.getVideoDevice();
+
+		for (int i = 0; i < cameras.length; i++) {
+			if (cameras[i].id == currentCameraId) {
+				final int newCameraId = i + 1 < cameras.length ? cameras[i + 1].id : cameras[0].id;
```
Comment (jritzerfeld) [inline app/src/main/java/org/simlar/service/liblinphone/LinphoneHandler.java]: This looks like a longer version of `final int newCameraId = cameras[(i + 1) % cameras.length]`.

File: app/src/main/java/org/simlar/service/liblinphone/LinphoneThread.java [hunk 4]
```diff
@@ -372,7 +549,17 @@ public void callState(final LinphoneCore lc, final LinphoneCall call, final Linp
...
-			Lg.i("callState changed state=", fixedState, " number=", new CallLogger(call), " message=", message);
+			final VideoState videoState = createVideoState(fixedState, call);
+
+			updateVideoState(videoState);
+
+			Lg.i("callState changed state=", fixedState, " number=", new CallLogger(call), " message=", message, " videoState=", videoState);
+
+			if (videoState == VideoState.REMOTE_REQUESTED) {
+				Lg.i("remote requested video");
+				/// NOTE: this needs to happen directly, posting to linphone thread might take to log
```
Comment (jritzerfeld) [inline app/src/main/java/org/simlar/service/liblinphone/LinphoneThread.java]: `... to long`?

File: app/src/main/java/org/simlar/service/liblinphone/LinphoneHandler.java (hunk 1 above)
Comment (BenSartor) [reply app/src/main/java/org/simlar/service/liblinphone/LinphoneHandler.java]: Honestly, I do not know :)
Linphone-Android does it that way. But I think, taking the first front facing camera is a good idea.

I have pushed a commit to the devel-video branch:
7780aa67286a5adf87096665e6b1666ec9be037d

File: app/src/main/java/org/simlar/service/liblinphone/LinphoneHandler.java (hunk 2 above)
Comment (BenSartor) [reply app/src/main/java/org/simlar/service/liblinphone/LinphoneHandler.java]: I think it is better to use another camera than. But warning makes sense for me:
edd0d8d1d5f84c438f82f3f293d494b16b80a261

And we should abort if there is no camera at all:
ed85f2719ff348c4676cc88c2a6f2f52dc6f8a89

File: app/src/main/java/org/simlar/service/liblinphone/LinphoneHandler.java (hunk 3 above)
Comment (BenSartor) [reply app/src/main/java/org/simlar/service/liblinphone/LinphoneHandler.java]: You are right and actually I already fixed this in the devel-video branch:
e3177b72d90b69897d4771cb37f32a05b4e8421a

File: app/src/main/java/org/simlar/widgets/CallActivity.java [hunk 5]
```diff
@@ -288,6 +319,151 @@ private void onCallConnectionDetailsChanged()
 		mConnectionDetailsDialogFragment.setCallConnectionDetails(mCommunicator.getService().getCallConnectionDetails());
 	}
 
+	private void onVideoStateChanged(final VideoState videoState)
+	{
+		Lg.i("onVideoStateChanged: ", videoState);
+
+		if (VideoState.PLAYING == videoState || VideoState.ENCRYPTING == videoState || VideoState.WAITING_FOR_ICE == videoState) {
```
[the remark this answers is not in the document]
Comment (BenSartor) [reply app/src/main/java/org/simlar/widgets/CallActivity.java]: You are right! And sorry, again I have changed it in the devel-video branch already:
[CallActivity.java:327](../blob/0bad01db1b5b807e08678980659154b7b736f6a4/app/src/main/java/org/simlar/widgets/CallActivity.java#L327)

File: app/src/main/java/org/simlar/service/liblinphone/LinphoneThread.java (hunk 4 above)
Comment (BenSartor) [reply app/src/main/java/org/simlar/service/liblinphone/LinphoneThread.java]: 0bad01db1b5b807e08678980659154b7b736f6a4

File: app/src/main/java/org/simlar/widgets/VideoFragment.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+		void onCaptureViewClick();
+	}
+
+	@Override
+	public final void onAttach(final Context context)
+	{
+		super.onAttach(context);
+		Lg.i("onAttach");
+
+		if ( ! (context instanceof Listener )) {
```
[the remark this answers is not in the document]
Comment (BenSartor) [reply app/src/main/java/org/simlar/widgets/VideoFragment.java]: 89ca6c29b0789263f93132be5f06a7cea8346641

**Pull request merged**